### PR TITLE
Adding usage doc for Natural Language API.

### PR DESCRIPTION
**Pull request by dhermes**
/cc @omaray are you in the org yet (I can't assign you to this PR)

### CODE REVIEW

File: docs/language-usage.rst [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+The Google Natural Language API has three supported methods
+
+- `analyzeEntities`_
+- `analyzeSentiment`_
+- `annotateText`_
+
+and each method uses a `Document`_ for representing text. To
+create a :class:`~gcloud.language.document.Document`,
+
+  .. code-block:: python
```
Comment (tseaver) [inline docs/language-usage.rst]: Why use doctest syntax inside `.. code-block:: python`?  It clutters stuff up, and makes it harder to cut-and-paste examples.

File: docs/language-usage.rst [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+     'PLAIN_TEXT'
+
+These values can also be specified in the constructor
+
+  .. code-block:: python
+
+     >>> html_content = """\
+     ... <html>
+     ...   <head>
+     ...     <title>Story Time</time>
```
Comment (tseaver) [inline docs/language-usage.rst]: El Tiempo de las Historias?  ;)

File: docs/language-usage.rst [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+specified in the request. By default, the UTF-8 is used:
+
+  .. code-block:: python
+
+     >>> document.analyze_entities(encoding=language.Encoding.UTF8)
+
+but the encoding can be one of
+:attr:`Document.UTF8 <gcloud.language.document.Encoding.UTF8>`,
+:attr:`Document.UTF16 <gcloud.language.document.Encoding.UTF16>`, or
+:attr:`Document.UTF32 <gcloud.language.document.Encoding.UTF32>`.
```
Comment (tseaver) [inline docs/language-usage.rst]: s/Document/Encoding/ for these three.

**Comment by tseaver** [comment]
A couple of more general questions:

- Shouldn't this PR (and subsequent implementation PRs) target a `natural-language-feature ` branch?
- Should we be packaging up the JSON results for the three APIs into objects?

File: docs/language-usage.rst [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+structure and meaning of text via powerful machine
+learning models. You can use it to extract information about
+people, places, events and much more, mentioned in text documents,
+news articles or blog posts. You can use it to understand
+sentiment about your product on social media or parse intent from
+customer conversations happening in a call center or a messaging
+app. You can analyze text uploaded in your request or integrate
+with your document storage on Google Cloud Storage.
+
+This is a Beta release of Google Cloud Natural Language API. This
```
Comment (daspecster) [inline docs/language-usage.rst]: Should this be a note or warning?

File: docs/language-usage.rst (hunk 1 above)
Comment (daspecster) [inline docs/language-usage.rst]: I should check the json-doc site for copy/pastability as well.

File: docs/language-usage.rst (hunk 4 above)
Comment (jgeewax) [inline docs/language-usage.rst]: Probably a warning.

File: docs/language-usage.rst (hunk 1 above)
Comment (jgeewax) [inline docs/language-usage.rst]: Yea. -1 on the REPL style here.

File: docs/language-usage.rst [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+- `analyzeSentiment`_
+- `annotateText`_
+
+and each method uses a `Document`_ for representing text. To
+create a :class:`~gcloud.language.document.Document`,
+
+  .. code-block:: python
+
+     >>> text_content = (
+     ...     'Google, headquartered in Mountain View, unveiled the '
```
Comment (jgeewax) [inline docs/language-usage.rst]: I know this is the example that we give on the website, but... I don't love it. Can we do one that's a bit more clear and less Google-fantastic?

File: docs/language-usage.rst [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+     ... <html>
+     ...   <head>
+     ...     <title>Story Time</time>
+     ...   </head>
+     ...   <body>
+     ...     <p>La vaca salt&oacute; sobre la luna.</p>
+     ...   </body>
+     ... </html>
+     ... """
+     >>> document = client.document(html_content,
```
Comment (jgeewax) [inline docs/language-usage.rst]: Is it crazy to have different document type factories here? ie, `Document.from_html()` and `Document.from_text()` ?

File: docs/language-usage.rst [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+     ...                            language='es',
+     ...                            doc_type=language.Document.HTML)
+
+The ``language`` arugment can be either ISO-639-1 or BCP-47 language
```
Comment (jgeewax) [inline docs/language-usage.rst]: Let's just say BCP-47. That's what Google will standardize one.

File: docs/language-usage.rst [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+codes; at the time, only English, Spanish, and Japanese are supported.
+However, the ``analyzeSentiment`` method only supports English text.
```
Comment (jgeewax) [inline docs/language-usage.rst]: Can we link off somewhere else for these validations? If the API adds support for new ones, we'll have to maintain this, which sucks. If we say "the list of supported languages is here" then we're in decent shape and the common case (English here), we're covered as people can assume it's there from examples.

File: docs/language-usage.rst [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+The document type (``doc_type``) argument can be one of
+:attr:`Document.PLAIN_TEXT <gcloud.language.document.Document.PLAIN_TEXT>` or
+:attr:`Document.HTML <gcloud.language.document.Document.HTML>`.
+
+In addition to supplying the text / HTML content, a document can refer
+to content stored in `Google Cloud Storage`_:
+
+  .. code-block:: python
+
+     >>> document = client.document(bucket='my-text-bucket',
```
Comment (jgeewax) [inline docs/language-usage.rst]: This seems like it'd lend itself even more to a factory for this (`Document.from_uri()` or `Document.from_blob(bucket=..., blob=...)`)

File: docs/language-usage.rst [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+.. _analyzeSentiment: https://cloud.google.com/natural-language/reference/rest/v1beta1/documents/analyzeSentiment
+.. _annotateText: https://cloud.google.com/natural-language/reference/rest/v1beta1/documents/annotateText
+.. _Document: https://cloud.google.com/natural-language/reference/rest/v1beta1/Document
+.. _Google Cloud Storage: https://cloud.google.com/storage/
+
+Analyze Entities
+----------------
+
+The :meth:`~gcloud.language.document.Document.analyze_entities` method
+finds named entities (i.e. proper names) in the text, entity types,
```
Comment (jgeewax) [inline docs/language-usage.rst]: can we put the examples in-line here? particularly for salience which often confuses people.

File: docs/language-usage.rst [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+salience (prominence), mentions for each entity, and other properties.
+
+  .. code-block:: python
+
+     >>> text_val = ("Michelangelo Caravaggio, Italian painter, is "
```
Comment (jgeewax) [inline docs/language-usage.rst]: text_content* ?

File: docs/language-usage.rst [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+     {
+       'entities': [
+         {
+           'name': 'Michelangelo Caravaggio',
+           'type': 'PERSON',
+           'metadata': {
+             'wikipedia_url': 'http://en.wikipedia.org/wiki/Caravaggio'
+           },
+           'salience': 0.75942981,
+           'mentions': [
```
Comment (jgeewax) [inline docs/language-usage.rst]: Can we do anything more concrete here? For example, `beginOffset` is in camel-case and `wikipedia_url` is snake.. this seems like we're not abstracting the data very much. Anything we can do to make tokenizing these things better?

File: docs/language-usage.rst [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+             {
+               'text': {
+                 'content': 'The Calling of Saint Matthew',
+                 'beginOffset': 56
+               }
+             }
+           ]
+         }
+       ],
+       'language': 'en'
```
Comment (jgeewax) [inline docs/language-usage.rst]: This isn't a BCP-47 code... ?

File: docs/language-usage.rst [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+     }
+
+In addition, the encoding of the text in the document can be
+specified in the request. By default, the UTF-8 is used:
+
+  .. code-block:: python
+
+     >>> document.analyze_entities(encoding=language.Encoding.UTF8)
```
Comment (jgeewax) [inline docs/language-usage.rst]: Why is this here? Shouldn't it be at document construction time? A document doesn't change encoding depending on the analysis you run...

File: docs/language-usage.rst [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+analyzes the sentiment of the provided text. Currently, this method
+only supports English text.
+
+  .. code-block:: python
+
+     >>> text_val = "Jogging isn't very fun."
+     >>> document = client.document(text_val)
+     >>> document.analyze_sentiment()
+     {
+         'documentSentiment': {
```
Comment (jgeewax) [inline docs/language-usage.rst]: Not a fan of this camel case and just returning the REST JSON output. :(

File: docs/language-usage.rst [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+The API can be used to extract any subset of `Features`_ among
+:attr:`~gcloud.language.document.Features.SYNTAX`,
+:attr:`~gcloud.language.document.Features.ENTITIES`, and
+:attr:`~gcloud.language.document.Features.DOCUMENT_SENTIMENT`.
+
+  .. code-block:: python
+
+     >>> text_val = 'The cow jumped over the moon.'
+     >>> document = client.document(text_val)
+     >>> features = [
```
Comment (jgeewax) [inline docs/language-usage.rst]: I worry about this being a bunch of magic variables that you pass in to a single annotate method (this is exactly what we *didn't* want to do in the Vision API with "annotate").

Can we use boolean variables to turn on/off these features? 

```python
document.annotate_text(syntax=True, sentiment=True, entities=False)
```

Comment (jgeewax) [inline docs/language-usage.rst]: Or break into separate methods for the various things?

```python
document.annotate_syntax()  # Or analyze... not sure which one it should be...
```

File: docs/language-usage.rst [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+     ...     language.Features.SYNTAX,
+     ...     language.Features.ENTITIES,
+     ...     language.Features.DOCUMENT_SENTIMENT,
+     ... ]
+     >>> document.annotate_text(features=features)
+     {
+         'tokens': [
```
Comment (jgeewax) [inline docs/language-usage.rst]: Can we abstract this any further? camel case here is fine for JSON, but we're supposed to let people use dot syntax to mess with stuff, not dictionary syntax.

Ie, 

```python
result = document.annotate_text(...)
tokens = result.tokens
nouns = filter(lambda t : return t.part_of_speech == language.Noun, tokens)
```

File: docs/language-usage.rst [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+                     'label': 'P'
+                 },
+                 'partOfSpeech': {
+                     'tag': 'PUNCT'
+                 },
+                 'lemma': '.'
+             }
+         ],
+         'entities': [],
+         'documentSentiment': {
```
Comment (jgeewax) [inline docs/language-usage.rst]: This is weird. Not sure why we have `language`, `tokens`, `entities` and then `documentSentiment` instead of just `sentiment`...

File: docs/language-usage.rst (hunk 6 above)
Comment (dhermes) [inline docs/language-usage.rst]: Overkill

File: docs/language-usage.rst (hunk 7 above)
Comment (dhermes) [inline docs/language-usage.rst]: This text is taken directly from the docs for the API

File: docs/language-usage.rst (hunk 12 above)
Comment (dhermes) [inline docs/language-usage.rst]: This is 0% abstracted, just the raw payload

File: docs/language-usage.rst (hunk 13 above)
Comment (dhermes) [inline docs/language-usage.rst]: It's what the API returns (see my comment above)

**Comment by dhermes** [comment]
@tseaver Can you weigh in here on which response values you think we should wrap?

**Comment by elibixby** [comment]
@jerjou Can you take a look to make sure we have someone who has used the API a bit to comment on usability.

**Comment by tseaver** [comment]
> Can you weigh in here on which response values you think we should wrap?

For `analyze_entities`, we could return instances of a (putative) `Entity` class:

```python
for entity in document.analyze_entities():
    build_entity_link(entity.name, entity.type, entity.url)
```


For `analyze_sentiment`, we could return a (putative) `Sentiment` instance:

```python
def by_sentiment(document):
    """Sort key:  rank from most negative to most positive."""
    sentiment = document.analyze_sentiment()
    return sentiment.polarity * sentiment.magnitude

ranked = sorted(documents, key=by_sentiment)
```

For `annotate_text`, (assuming we have flags to disable unwanted features):

```python
only_syntax = document.annotate_text(
    include_entities=False, include_sentiment=False)

assert only_syntax.entities is None
assert only_syntax.sentiment is None

tree = ParseTree()  # imagined class w/ 'add_token' method.

for index, token in enumerate(only_syntax.tokens):
    edge = token.dependency_edge
    tree.add_token(
        index, token.lemma, token.part_of_speech, edge.head_index, edge.label)
```

File: docs/language-usage.rst (hunk 16 above)
Comment (tseaver) [inline docs/language-usage.rst]: I would have the flags default to False, and have the method return an instance with appropriate attributes, set to `None` if not present in the response.

File: docs/language-usage.rst (hunk 9 above)
Comment (tseaver) [inline docs/language-usage.rst]: +1 for a `Document.from_blob()` classmethod factory.

File: docs/language-usage.rst [hunk 19]
```diff
@@ -0,0 +1,354 @@
...
+
+     >>> text_content = (
+     ...     'Google, headquartered in Mountain View, unveiled the '
+     ...     'new Android phone at the Consumer Electronic Show.  '
+     ...     'Sundar Pichai said in his keynote that users love '
+     ...     'their new Android phones.')
+     >>> document = client.document(text_content)
+
+By default, the document's language and type are English and
+plain text:
```
Comment (jerjou) [inline docs/language-usage.rst]: Could we instead infer [language](http://stackoverflow.com/questions/3425294/how-to-detect-the-os-default-language-in-python) from the environment? Presumably we want to make this convenient for all users in their native language?

File: docs/language-usage.rst (hunk 9 above)
Comment (jerjou) [inline docs/language-usage.rst]: +1 ability to specify uri instead of bucket/object pair

File: docs/language-usage.rst [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+For an overview of authentication in ``gcloud-python``, see
+:doc:`gcloud-auth`.
+
+Assuming your environment is set up as described in that document,
+create an instance of :class:`~gcloud.language.client.Client`.
+
+  .. code-block:: python
+
+     >>> from gcloud import language
+     >>> client = language.Client()
```
Comment (tseaver) [inline docs/language-usage.rst]: Should the name reflect the back-end API name?  E.g.:

```python
from gcloud import natural_language
client = natural_language.Client()
```

File: docs/language-usage.rst [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+               }
+             }
+           ]
+         }
+       ],
+       'language': 'en'
+     }
+
+In addition, the encoding of the text in the document can be
+specified in the request. By default, the UTF-8 is used:
```
Comment (jerjou) [inline docs/language-usage.rst]: For python, we should default to UTF16 or UTF32, depending on [sys.maxunicode](http://stackoverflow.com/questions/13095499/unicode-in-python-just-utf-16).

File: docs/language-usage.rst (hunk 20 above)
Comment (tseaver) [inline docs/language-usage.rst]: Should the `Client` take some sensibly-defaulted parameters, e.g., `encoding`, `language`, and then supply those values as defaults for other methods?

File: docs/language-usage.rst (hunk 19 above)
Comment (tseaver) [inline docs/language-usage.rst]: @jerjou As the second answer to that SO post notes, guessing is hard to get right.

File: docs/language-usage.rst (hunk 12 above)
Comment (jerjou) [inline docs/language-usage.rst]: It'd be nice if it was >0% abstracted :-)
Normalizing the casing of the response keys seems to be within gcloud-python's mission to make REST apis feel more pythonic, no?

File: docs/language-usage.rst (hunk 9 above)
Comment (tseaver) [inline docs/language-usage.rst]: @jerjou What is the format that the back-end expects for the URI?  The [`Document` spec](https://cloud.google.com/natural-language/reference/rest/v1beta1/Document) doesn't say.

-  Bigquery uses `gs://<bucket_name>/<object_name_or_glob>` for its [source URIs](https://cloud.google.com/bigquery/docs/loading-data-cloud-storage).
- Logging uses `storage.googleapis.com/<bucket>` for its `LogSink` [Cloud Storage destination URIs](https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/projects.sinks#LogSink.FIELDS.destination).

It might be more natural to create a `Blob` instance and hand it to the `Document` factory, which would mean that the user didn't need to care about the URI formats.

File: docs/language-usage.rst (hunk 19 above)
Comment (jerjou) [inline docs/language-usage.rst]: Ah - I guess I should learn to read more than the first answer 😅
I still think it'd have value as a default value, though I can also see how that might make things unnecessarily complicated (especially since the API currently only supports a handful of languages).

File: docs/language-usage.rst (hunk 9 above)
Comment (jerjou) [inline docs/language-usage.rst]: It's the gs://bucket/object uri. We use that uri format fairly consistently across our apis (eg the [vision api](https://cloud.google.com/vision/reference/rest/v1/images/annotate#ImageSource), [speech api](https://cloud.google.com/speech/reference/rest/v1beta1/RecognitionAudio), [gsutil](https://cloud.google.com/storage/docs/gsutil), [dataflow](https://cloud.google.com/dataflow/java-sdk/JavaDoc/com/google/cloud/dataflow/sdk/io/AvroIO)). I dunno what's going on with logging 😶

File: docs/language-usage.rst (hunk 21 above)
Comment (tseaver) [inline docs/language-usage.rst]: Hmm, I've actually never used either of those codecs in 18 years of Python development, including workiong with [multilingual sites](http://www.rfa.org/english/). UTF-8 is pretty dominant on the web except in the Big 5 world.

If we are talking about i user-supplied bytes, I don't know whether there *is* a sensible default:  UTF-16 is the natural encoding for Java / Jython, and maybe for Windows APIs, but the bytes are in whatever encoding the user (or her tool) chose when saving the file.

Comment (jerjou) [inline docs/language-usage.rst]: So what's happening here is somewhat subtle. The string that's sent across the wire, I believe, is encoded as UTF8 regardless of the encoding the language uses. However, the API needs to calculate offsets into the string for the entities/etc it detects, which will differ depending on the codec. For example, given the string `1😶2`, in UTF8, the index for `2` is 5. In UTF16, it's 3. In UTF32, it's 2. This is because the index into the string is an index into an array of entries of the smallest 'character' size - 8-bits for UTF8, 16-bits for UTF16, 32-bits for UTF32. So double-byte characters will take up two slots in a UTF8-encoded string, and one slot in a UTF16-encoded string; and triple-byte characters will take up three slots in a UTF8-encoded string, and two slots in a UTF16-encoded string; etc.

So the API needs to know what encoding to calculate the offsets for.

python2 `string` objects are basically byte arrays that are encoded in ascii by default. This barfs on unicode:

```python
>>> a = u'1😶2'
>>> str(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\U0001f636' in position 1: ordinal not in range(128)
```

You can explicitly specify the encoding for unicode characters, for example to UTF8:

```python
>>> a.encode('utf8')
'1\xf0\x9f\x98\xb62'
>>> a.encode('utf8')[5]
'2'
```
in which case you would pass UTF8 to the API.

But if you're using `unicode` objects (ie not explicitly using a particular encoding), you get either UTF32 or UTF16:

```python
>>> sys.maxunicode
1114111
>>> a[2]
u'2'
```
depending on what your python interpreter is compiled as.

Does that make sense? I go through a bunch of examples [here](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/language/api/analyze_test.py#L63) if that would help.

Comment (jerjou) [inline docs/language-usage.rst]: But yes - you're correct in that if you get an arbitrary byte string from the user, utf8 is as sensible a default as any, since it's so ubiquitous. But if you get an arbitrary `unicode` string (or if you're working in python3, where I believe everything is a `unicode` by default), then it makes sense to default to UTF16/32.

Comment (tseaver) [inline docs/language-usage.rst]: I think I'm misunderstanding something:  Given that `natural-language` is a JSON-over-HTTP only API, we are going to be encoding user-supplied Unicode on the wire as UTF-8.  What would be the point of having the back-end return offsets based on arbitrary larger code-point sizes?  The "logical" return value for those offsets would be to index the code points (the actual Unicode characters after decoding the wire-transmitted UTF-8), rather than a purely arbitrary number of bytes into an encoded version the user never contemplated.

Comment (jerjou) [inline docs/language-usage.rst]: Right. The indexes aren't into the JSON-encoded utf8 byte string - they're into a string encoded in the specified encoding. A python unicode object is encoded in UTF32  (assuming the python interpreter was compiled in wide mode), which is how each code point is one index. So by specifying `encoding=UTF32`, the returned offsets will correspond to the python `unicode` object, which treats every code point as one index (rather than, say, UTF16, where triple-byte characters will occupy two indices). The reason I'm suggesting UTF32 be the default is _so that_ the user doesn't have to think about the encoding, since their string is probably already UTF32 (assuming they're using unicode objects, rather than python2 strings).

Then there's the issue that, depending on whether your python interpreter was compiled in [wide or narrow mode](http://stackoverflow.com/questions/13095499/unicode-in-python-just-utf-16), the unicode object might actually be UTF16.

Also note, if you're writing an app where you eventually pass the offsets to javascript (which encodes in UTF16), you'll need the offsets for UTF16 rather than UTF32. It all depends on what string you're going to be applying the offsets to. But if it's going to be applied to python `unicode` objects, it should be either UTF32 or UTF16, depending on `sys.maxunicode`

File: docs/language-usage.rst (hunk 4 above)
Comment (dhermes) [inline docs/language-usage.rst]: FYI this is copy-pasted from the actual backend API docs. Will put it in a warning section

File: docs/language-usage.rst (hunk 20 above)
Comment (dhermes) [inline docs/language-usage.rst]: I am fine with that. Adding now.

File: docs/language-usage.rst (hunk 5 above)
Comment (dhermes) [inline docs/language-usage.rst]: Any suggestions?

File: docs/language-usage.rst (hunk 19 above)
Comment (dhermes) [inline docs/language-usage.rst]: Agree with @tseaver. Detecting language would be biting off more than we can chew.

File: docs/language-usage.rst (hunk 2 above)
Comment (dhermes) [inline docs/language-usage.rst]: Doing it

File: docs/language-usage.rst (hunk 7 above)
Comment (dhermes) [inline docs/language-usage.rst]: @jgeewax Should we contradict / leave out info from the official docs?

File: docs/language-usage.rst (hunk 8 above)
Comment (dhermes) [inline docs/language-usage.rst]: I came up with https://cloud.google.com/natural-language/docs/ for a link that shows only `en`, `es`, `ja` and https://cloud.google.com/natural-language/reference/rest/v1beta1/documents/analyzeSentiment#body.request_body.FIELDS.document as a link that says only `en` for sentiment.

The first isn't a very clear marker of where the information is, but it just isn't in the page in an easy to link to way.

File: docs/language-usage.rst (hunk 6 above)
Comment (dhermes) [inline docs/language-usage.rst]: Also, if we want to use a factory from the `client`, then the actual `Document` class isn't "useful" (in some sense)

File: docs/language-usage.rst (hunk 11 above)
Comment (dhermes) [inline docs/language-usage.rst]: Done.

File: docs/language-usage.rst (hunk 10 above)
Comment (dhermes) [inline docs/language-usage.rst]: I'm not sure what you mean.

File: docs/language-usage.rst (hunk 14 above)
Comment (dhermes) [inline docs/language-usage.rst]: :+1: On it

File: docs/language-usage.rst (hunk 3 above)
Comment (dhermes) [inline docs/language-usage.rst]: Thanks.

File: docs/language-usage.rst (hunk 15 above)
Comment (dhermes) [inline docs/language-usage.rst]: Yup. Wrapping all the return values to make them python.

File: docs/language-usage.rst (hunk 16 above)
Comment (dhermes) [inline docs/language-usage.rst]: Going the route suggested by @tseaver

File: docs/language-usage.rst (hunk 17 above)
Comment (dhermes) [inline docs/language-usage.rst]: As above, this was just the JSON payload returned by the API. As suggested by @tseaver I am wrapping them in native object.

>  we're supposed to let people use dot syntax to mess with stuff, not dictionary syntax.

"supposed to" since when?

File: docs/language-usage.rst [hunk 22]
```diff
@@ -35,6 +37,19 @@ create an instance of :class:`~gcloud.language.client.Client`.
      >>> from gcloud import language
      >>> client = language.Client()
 
+By default the ``language`` is ``'en'`` and the ``encoding`` is
+UFT-8. To over-ride these values:
```
Comment (tseaver) [inline docs/language-usage.rst]: ``UTF-8``

Comment (dhermes) [inline docs/language-usage.rst]: OOF

File: docs/language-usage.rst [hunk 23]
```diff
@@ -112,243 +158,122 @@ Analyze Entities
...
-need in-depth text features to build upon.
-
-The API can be used to extract any subset of `Features`_ among
-:attr:`~gcloud.language.document.Features.SYNTAX`,
-:attr:`~gcloud.language.document.Features.ENTITIES`, and
-:attr:`~gcloud.language.document.Features.DOCUMENT_SENTIMENT`.
+analyzes a document and is intended for users who are familiar with
+machine learning and need in-depth text features to build upon.
+
+The method returns a four-tuple of
```
Comment (tseaver) [inline docs/language-usage.rst]: Use a namedtuple, to allow access by attr?

Comment (dhermes) [inline docs/language-usage.rst]: SGTM

**Comment by tseaver** [comment]
LGTM.

**Comment by dhermes** [comment]
@jgeewax In the interest of keeping this PR from reaching 15 days of age, I'll merge with @tseaver's LGTM and we can address after the fact as the API develops (it's a small surface area so it won't be overwhelming).

File: docs/language-usage.rst [hunk 24]
```diff
@@ -0,0 +1,279 @@
...
+     >>> from gcloud import language
+     >>> client = language.Client()
+
+By default the ``language`` is ``'en'`` and the ``encoding`` is
+UTF-8. To over-ride these values:
+
+  .. code-block:: python
+
+     >>> client = language.Client(language='es',
+     ...                          encoding=encoding=language.Encoding.UTF16)
```
Comment (jgeewax) [inline docs/language-usage.rst]: encoding* ?

Comment (dhermes) [inline docs/language-usage.rst]: oy

Comment (dhermes) [inline docs/language-usage.rst]: Thanks

File: docs/language-usage.rst [hunk 25]
```diff
@@ -0,0 +1,279 @@
...
+create a :class:`~gcloud.language.document.Document`,
+
+  .. code-block:: python
+
+     >>> text_content = (
+     ...     'Google, headquartered in Mountain View, unveiled the '
+     ...     'new Android phone at the Consumer Electronic Show.  '
+     ...     'Sundar Pichai said in his keynote that users love '
+     ...     'their new Android phones.')
+     >>> document = client.document_from_text(text_content)
```
Comment (jgeewax) [inline docs/language-usage.rst]: Could this pattern be `client.document.from_text()` ? Or `client.text_document(...)` and `client.html_document()` ?

File: docs/language-usage.rst [hunk 26]
```diff
@@ -0,0 +1,279 @@
...
+     >>> document.doc_type == language.Document.PLAIN_TEXT
+     True
+
+In addition, the document's language defaults to the language on
+the client
+
+  .. code-block:: python
+
+     >>> document.language
+     'en'
```
Comment (jgeewax) [inline docs/language-usage.rst]: can we default to `en-US` (the BCP-47 code) which is what we should be getting back?

File: docs/language-usage.rst [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+     ...   </head>
+     ...   <body>
+     ...     <p>La vaca salt&oacute; sobre la luna.</p>
+     ...   </body>
+     ... </html>
+     ... """
+     >>> document = client.document_from_html(html_content,
+     ...                                      language='es')
+
+The ``language`` argument can be either ISO-639-1 or BCP-47 language
```
Comment (jgeewax) [inline docs/language-usage.rst]: Can we focus on BCP-47 rather than ISO-639-1 (that's the direction the APIs will be going)

File: docs/language-usage.rst [hunk 28]
```diff
@@ -0,0 +1,279 @@
...
+:attr:`Document.PLAIN_TEXT <gcloud.language.document.Document.PLAIN_TEXT>` or
+:attr:`Document.HTML <gcloud.language.document.Document.HTML>`.
+
+In addition to supplying the text / HTML content, a document can refer
+to content stored in `Google Cloud Storage`_. We can use the
+:meth:`~gcloud.language.client.Client.document_from_blob` method:
+
+  .. code-block:: python
+
+     >>> document = client.document_from_blob(bucket='my-text-bucket',
```
Comment (jgeewax) [inline docs/language-usage.rst]: I think this one might be overkill -- I think @tseaver mentioned that also. The goal of `from_document` (or the comment above's mention of `html_document` was about not typing `type=language.Document.PLAIN_TEXT` as a parameter rather than using difference sources (blob vs uri).

I'd expect `from_blob()` to take a `Blob` object (`google.cloud.storage.blob.Blob`) so that code would look like

```python
blob = <somehow get a blob>
doc = client.document_from_blob(blob)
```

rather than

```python
blob = <some blob>
doc = client.document_from_blob(blob.bucket, blob.name)
```

File: docs/language-usage.rst [hunk 29]
```diff
@@ -0,0 +1,279 @@
...
+     True
+
+and the :meth:`~gcloud.language.client.Client.document_from_uri`
+method. In either case, the document type can be specified with
+the ``doc_type`` argument:
+
+  .. code-block:: python
+
+     >>> gcs_url = 'gs://my-text-bucket/sentiment-me.txt'
+     >>> document = client.document_from_uri(
```
Comment (jgeewax) [inline docs/language-usage.rst]: Does this work by pulling locally and sending the content? or by sending Google the url to load remotely? Can we document that?

File: docs/language-usage.rst [hunk 30]
```diff
@@ -0,0 +1,279 @@
...
+     >>> document.doc_type == language.Document.PLAIN_TEXT
+     True
+
+and the :meth:`~gcloud.language.client.Client.document_from_uri`
+method. In either case, the document type can be specified with
+the ``doc_type`` argument:
+
+  .. code-block:: python
+
+     >>> gcs_url = 'gs://my-text-bucket/sentiment-me.txt'
```
Comment (jgeewax) [inline docs/language-usage.rst]: Can we auto-detect HTML based on the content-type or file extension (`.html`, `.htm`) ? If we do that, then life gets a bit easier on these, right?

File: docs/language-usage.rst [hunk 31]
```diff
@@ -0,0 +1,279 @@
...
+The :meth:`~gcloud.language.document.Document.analyze_sentiment` method
+analyzes the sentiment of the provided text and returns a
+:class:`~gcloud.language.sentiment.Sentiment`. Currently, this method
+only supports English text.
+
+  .. code-block:: python
+
+     >>> text_content = "Jogging isn't very fun."
+     >>> document = client.document(text_content)
+     >>> sentiment = document.analyze_sentiment()
```
Comment (jgeewax) [inline docs/language-usage.rst]: is this stored statefully on the document at all? ie, can i make this call and have it cached for next time?

File: docs/language-usage.rst [hunk 32]
```diff
@@ -0,0 +1,279 @@
...
+     >>> print(sentiment.polarity)
+     -1
+     >>> print(sentiment.magnitude)
+     0.8
+
+Annotate Text
```
Comment (jgeewax) [inline docs/language-usage.rst]: Should this be called annotate text when the others are analyze_thing? Seems like this is just the entire thing (analyze_everything). Would .analyze() make sense?

File: docs/language-usage.rst (hunk 25 above)
Comment (dhermes) [inline docs/language-usage.rst]: Filed #2183

File: docs/language-usage.rst (hunk 26 above)
Comment (dhermes) [inline docs/language-usage.rst]: Changed to `en-US`. The return value just echoes the one passed in, but if none if passed, the default is `en`

File: docs/language-usage.rst (hunk 27 above)
Comment (dhermes) [inline docs/language-usage.rst]: Filed #2184

File: docs/language-usage.rst (hunk 28 above)
Comment (dhermes) [inline docs/language-usage.rst]: Cool. Will remove. Filed #2173 for the actual `Blob` use case

File: docs/language-usage.rst (hunk 29 above)
Comment (dhermes) [inline docs/language-usage.rst]: It's a feature of the API. Either `content` is sent in the API request or `gcsContentUri` is. The backend loads the content for us, we just send the name of the doc.

We are essentially re-documenting the backend API by doing that (which is fine). Where to put it?

File: docs/language-usage.rst (hunk 30 above)
Comment (dhermes) [inline docs/language-usage.rst]: Sure. Filed #2185. Will follow up.

File: docs/language-usage.rst (hunk 31 above)
Comment (dhermes) [inline docs/language-usage.rst]: That would be the first ever cached result re-used for future API calls for this library. Not really something to just introduce haphazardly without a larger discussion.

File: docs/language-usage.rst (hunk 32 above)
Comment (dhermes) [inline docs/language-usage.rst]: These are the literal names of the backend methods. We can name whatever we like, and we have changed names at times to make the methods easier to grok. In this case, I thought the backend names were easy enough to understand that it wasn't worth inventing our own vocubalary that would differ from the backend docs.

Happy to change it to `.analyze()` if you like.
